### PR TITLE
[E92] Used sets as lookup tables to remember chains

### PR DESCRIPTION
**Pull request by qwrtln**

### CODE REVIEW

File: 92.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
+def add_digits_squared(n):
+    result = 0
+    while n:
+        result += (n % 10) ** 2
+        n /= 10
+    return result
+
+lookup89 = set()
+lookup1 = set()
```
Comment (zslm) [inline 92.py]: You can include 1 and 89 in the relevant lookup sets straight away. This way you also don't need to check for j == 1 and j == 89 below:

lookup89 = {89}
lookup1 = {1}

File: 92.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
+def add_digits_squared(n):
```
Comment (qwrtln) [inline 92.py]: Recursion is significantly slower in this case:

With recursion:
Exec time: 19.2057828903 [s]
Exec time: 19.7803030014 [s]
Exec time: 20.0441379547 [s]

With function as is:
Exec time: 14.9110360146 [s]
Exec time: 14.8208310604 [s]
Exec time: 15.6257121563 [s]

**Comment by qwrtln** [comment]
I have no idea what you mean by that but I think the answer is yes :-)

**Comment by qwrtln** [comment]
Yes, I can create a more efficient lookup table (with 567 beeing the upper limit). Exect time was reduced to ~11 sec

File: 92.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        return True
+    elif j == 89:
+        lookup[i] = 89
+        return True
+    return False
+
+# Highest digit sum comes from seven nines (10**7 - 1)
+limit = 9**2 * 7 + 1
+
+lookup = {}
```
Comment (zslm) [inline 92.py]: There are a lot of repeated final_sum calls for the same j. Try that:

lookup = {1: 1, 89: 89}

def final_sum(i):
  if i in lookup:
    return lookup[i]
  ret = final_sum(add_digits_squared(i))
  lookup[i] = ret
  return ret

for i in xrange(1, limit):
  lookup[i] = final_sum(i)

Comment (qwrtln) [inline 92.py]: Son of a... :-) I knew about the repetitions but I had no idea they could be handled that simply. Squares are also cached.

Exec time: 10.1895830631
Exec time: 10.2718720436
Exec time: 10.1934328079

**Pull request merged**